libyul/AsmAnalysis.cpp

```diff
@@ -641,45 +641,47 @@ void AsmAnalyzer::warnOnInstructions(solidity::Instruction _instr, SourceLocatio
 	solAssert(m_evmVersion.hasBitwiseShifting() == m_evmVersion.hasCreate2(), "");
 	solAssert(m_dialect->flavour != AsmFlavour::Yul, "");
 
-	if (_instr == solidity::Instruction::EXTCODEHASH)
+	auto warningForVM = [=](string const& vmKindMessage) {
 		m_errorReporter.warning(
 			_location,
 			"The \"" +
 			boost::to_lower_copy(instructionInfo(_instr).name)
-			+ "\" instruction is not supported by the VM version \"" +
-			"" + m_evmVersion.name() +
-			"\" you are currently compiling for. " +
-			"It will be interpreted as an invalid instruction on this VM."
-		);
-	else if ((
-		_instr == solidity::Instruction::RETURNDATACOPY ||
-		_instr == solidity::Instruction::RETURNDATASIZE ||
-		_instr == solidity::Instruction::STATICCALL
-	) && !m_evmVersion.supportsReturndata())
-		m_errorReporter.warning(
-			_location,
-			"The \"" +
-			boost::to_lower_copy(instructionInfo(_instr).name)
-			+ "\" instruction is only available for Byzantium-compatible VMs. " +
+			+ "\" instruction is " +
+			vmKindMessage +
+			" VMs. " +
 			"You are currently compiling for \"" +
 			m_evmVersion.name() +
 			"\", where it will be interpreted as an invalid instruction."
 		);
+	};
+
+	if ((
+		_instr == solidity::Instruction::RETURNDATACOPY ||
+		_instr == solidity::Instruction::RETURNDATASIZE
+	) && !m_evmVersion.supportsReturndata())
+	{
+		warningForVM("only available for Byzantium-compatible");
+	}
+	else if (_instr == solidity::Instruction::STATICCALL && !m_evmVersion.hasStaticCall())
+	{
+		warningForVM("only available for Byzantium-compatible");
+	}
 	else if ((
 		_instr == solidity::Instruction::SHL ||
 		_instr == solidity::Instruction::SHR ||
-		_instr == solidity::Instruction::SAR ||
-		_instr == solidity::Instruction::CREATE2
+		_instr == solidity::Instruction::SAR
 	) && !m_evmVersion.hasBitwiseShifting())
-		m_errorReporter.warning(
-			_location,
-			"The \"" +
-			boost::to_lower_copy(instructionInfo(_instr).name)
-			+ "\" instruction is only available for Constantinople-compatible VMs. " +
-			"You are currently compiling for \"" +
-			m_evmVersion.name() +
-			"\", where it will be interpreted as an invalid instruction."
-		);
+	{
+		warningForVM("only available for Constantinople-compatible");
+	}
+	else if (_instr == solidity::Instruction::CREATE2 && !m_evmVersion.hasCreate2())
+	{
```

Comment:
Could remove all those pesky curly braces.

Reply:
I've added them deliberately because it was hard to read the source otherwise (the if conditions are quite big).